### ISSUE TITLE
Rule: Require assigning `this` to a variable

### ISSUE THREAD

**Issue by jacksonrayhamilton** [eslint/eslint#5351]
**When does this rule warn? Please describe and show example code:**

This rule would only permit the `this` keyword if `this` were being assigned to a variable.

The following patterns are considered problems:

```js
function A() {
  this.b = 0; // `this` not assigned to a variable (for consistency)
  this.c = function () {
    this.b = 1; // `this` should reference caller, but will fail if method is detached
  };
  // The next 4 lines are okay, but demonstrate how `this` is usually only assigned to a
  // variable ad-hoc, rather than consistently
  var a = this;
  var d = function () {
    a.b = 2;
  };
  var e = function () {
    this.b = 3;
  }.bind(this); // Binding with `.bind` is easy to forget
  function f() {
    this.h = 4; // It is unclear which `this` is being referred-to
  }
  var i = () => this; // It is unclear which `this` is being referred-to
}
A.prototype.j = function () {
  this.b = 5; // `this` should reference caller, but will fail if method is detached
};
```

The following patterns are not considered problems:

```js
function A() {
  var a = this;
  a.b = 0; // Good; referenced via a variable (for consistency)
  a.c = function () {
    a.b = 1; // Good; binding cannot be lost
  };
  var d = function () {
    a.b = 2;
  };
  var e = function () {
    a.b = 3; // Good; this private function doesn't need to remember `.bind`
  };
  function f() {
    var g = this; // Good; it's obvious this is a different `this` than the outer `this`
    g.h = 4;
  }
  // Good; it is obvious we want to refer to `a`, and that we aren't just using
  // arrows because they are shorter than `function`
  var i = () => a;
}
A.prototype.j = function () {
  // Binding could still be lost if method is detached, but this is still worthwhile
  // for the other reasons
  var a = this;
  a.b = 5;
};
```

**Is this rule preventing an error or is it stylistic?**

This rule prevents errors. `this` is always originally obtained dynamically, but to use it in nested functions, we'd like it to be available lexically. Requiring `this` be assigned to a variable ensures lexical access, and can differentiate an outer `this` from a different inner `this`.

Arrow functions cannot obtain dynamic `this` values. Attempting to use arrow functions to bind `this` could cause confusion and/or unexpected results if a function could also have a dynamic `this` value. Therefore, this rule prevents errors in JavaScript environments both old and new.

Requiring `this` to *always* be assigned to a variable (even if there are no inner functions, or no different inner `this` values) also eases refactoring, as it is possible `this` will eventually need to be assigned anyway.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

JavaScript is an object-oriented language. `this` binding errors are endemic to object-oriented JavaScript. This rule prevents such errors. Therefore, it would be useful for many developers.

This technique is documented in _JavaScript: The Good Parts_. I have seen `this` assigned to variables in many projects, albeit usually only "as necessary." I don't know if it's possible to statically determine if assignment is "necessary," but consistently assigning `this` is probably better anyway.

**Are you willing to create the rule yourself?**

Yes.

**Comment by BYK** [comment]
I think with the introduction of arrow functions and `.bind` makes this an anti-pattern for :-1: for including this in core. That said this is a good idea for a custom plugin if you want to use it.

Btw. I think you should trim down your examples a bit to make it very clear which part would cause a warning and would not. Inline comments or just shorter examples can help.

**Comment by mysticatea** [comment]
Thank you for this issue.

I think `no-invalid-this` rule has the same purpose. 
Is `no-invalid-this` rule not enough to you?

**Comment by jacksonrayhamilton** [comment]
> I think arrow functions and `.bind` makes this an anti-pattern so :-1: for including this in core.

Arrow functions do not completely (adequately) solve `this` binding issues. Actually, I contend that arrow functions and `.bind` are both anti-patterns. Assigning `this` to a variable solves the same issues in a more consistent manner.

For arrows, see my post [here](http://stackoverflow.com/a/28135120/1468130), under "Regarding lexical `this`". You're free to disagree, but at the very least this would represent one view among the diverse set of viewpoints on how to effectively bind `this` in new ECMAScript versions.

As for ES5 and `.bind`: At one point I wrote code with lots of nested callbacks, and tried really hard to use `.bind` on every single one of them. Then, I forgot one. Now I don't use `.bind` any more, because I realized that it is too hard to remember to use it on every function. It is much easier to remember to assign `this` to a variable once at the top of a function.

> I think `no-invalid-this` rule has the same purpose. 
> Is `no-invalid-this` rule not enough to you?

`no-invalid-this` helps to mitigate the possibility of `this` binding errors, and could be used in concert with this rule. But `no-invalid-this` is not sufficient, because it tries to enable one to use `ThisExpression` everywhere, which is a demonstrably problematic practice.

**Comment by BYK** [comment]
I'd prefer you to not express your own opinions as mere facts. Arrow functions and `.bind` fixes the issues you mentioned quite properly. jQuery or any other library overloading `this` is not the fault of those features.

Anyways, I'm not convinced but I'm just one person so deferring to @eslint/eslint-team for more opinions.

**Comment by jacksonrayhamilton** [comment]
Apologies, I'm not trying to express opinion as fact. When I use language like "it is too hard" and "it is much easier," I actually do mean "for myself;" I use that language with the expectation that it would be true for other people (just based on counting the number of places where an error could occur).

FWIW, ESLint does take a library into account in one of [its examples](http://eslint.org/docs/rules/consistent-this), which seems to teeter on inconsistency with [the "library agnostic" guideline](http://eslint.org/docs/developer-guide/contributing/new-rules#core-rule-guidelines); however, it does reveal that ESLint is sometimes willing to be pragmatic w.r.t. the conventions of libraries.

**Comment by BYK** [comment]
@jacksonrayhamilton no worries, I just thought your response a bit too aggressive :)

Anyways, your points make sense, hence I deferred back to the team. Me not being convinced doesn't mean it won't end up in ESLint core ;)

**Comment by nzakas** [comment]
@jacksonrayhamilton Thanks for the suggestion. While you bring up some interesting points, I think this is best left as a custom rule. It's a bit too specific to a pattern that I haven't seen a lot of. The good news is that you can create your own rules and publish them as plugins, so you can still make this rule for yourself: http://eslint.org/docs/developer-guide/working-with-plugins#rules-in-plugins

**Comment by jacksonrayhamilton** [comment]
> It's a bit too specific to a pattern that I haven't seen a lot of.

That's unfortunate for me to hear. A reason I would like for this functionality to be built-in would be to raise awareness about the continued effectiveness of assigning `this`. I don't find the trade-offs of the newer binding methods worthwhile, despite that they seem to be popular.

I have another idea, which might result in less specificity: We could update `no-invalid-this` to be a bit more general, while maintaining the spirit of the rule's name. The viable cases for `no-invalid-this` could be customized, and an additional case could be offered. Right now, that rule requires `this` to appear in classes, class-like objects, and bound functions. We could change the rule so that the user specifies which locations are appropriate for `this` to appear, and add one more location: "Variable assignment."

The default setting (for backwards-compatibility) could be `{"classes": true, "class-like objects": true, "bound functions": true}`. With the new option, "variable assignment," we could ensure `this` is only used when assigning it to a variable; so, to obtain the functionality I desire, I could specify `{"variable assignment": true}`. I could optionally specify the previous options, as well, to further restrict myself to variable assignments inside classes, class-like objects and bound functions.

Do you think this would be a better route for integrating this functionality?

**Comment by platinumazure** [comment]
@jacksonrayhamilton Not speaking for the core team, but I think you would do better to "show me the money" by creating a custom rule in a plugin, and then let the NPM downloads prove the utility of your use case. At this point, I (at least) am skeptical, and I think you would need to evangelize your suggested pattern (but that the issue tracker of this repo is not the place to do it). If you believe strongly, you would probably do well to make your case in a blog as well, where you can more clearly show examples and counterexamples for your proposed style. I doubt you'll make any more headway here, in any case (but the core team may prove me wrong).

**Comment by xjamundx** [comment]
Happy to help anyone write a custom rule and publish it to NPM.. Always a
great testing ground for new ideas! You at all interested
@jacksonrayhamilton?

**Comment by jacksonrayhamilton** [comment]
@platinumazure, I have been trying to get into blogging for a while, I think that would be a good outlet to share my findings in code quality and style.

Thanks for the offer @xjamundx! I am pretty comfortable with ASTs and am confident I can write this myself.
